File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.swift
import Foundation

func getCalculate(_ result : Int, _ idx : Int) {
    if idx == N {
        result_max = max(result, result_max)
        result_min = min(result, result_min)
        return
    }
    for i in 0...3 {
        if op[i] > 0 {
            op[i] -= 1
            switch i {
                case 0 :
                    getCalculate(result + lst[idx], idx + 1)
                    op[i] += 1
                case 1 :
                    getCalculate(result - lst[idx], idx + 1)
                    op[i] += 1
                case 2 :
                    getCalculate(result * lst[idx], idx + 1)
                    op[i] += 1
                default:
                    if result < 0 {
                        getCalculate((abs(result) / lst[idx])*(-1), idx + 1)
                        op[i] += 1
                    }else{
                        getCalculate(result / lst[idx], idx + 1)
                        op[i] += 1
                    }
            }
        }
    }
}
let N = Int(readLine()!)! 
let lst : [Int] = readLine()!.split(separator: " ").map{Int(String($0))!} 
var op : [Int] = readLine()!.split(separator: " ").map{Int(String($0))!} 
var opList : [Character] = [Character]()

var result_max = -1000000001
var result_min = 1000000001
getCalculate(lst[0], 1)
print(result_max)
print(result_min)

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashMap;

public class Main {
    public static int ops[], selectedOps[], arr[], maxValue, minValue, N;
    public static boolean[] selected;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        arr = new int[N];
        arr = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int[] op = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        ops = new int[N - 1];
        selected = new boolean[ops.length];
        selectedOps = new int[N - 1];
        maxValue = -1000000000;
        minValue = 1000000000;
        int idx = 0;
        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < op[i]; j++) {
                ops[idx++] = i;
            }
        }

        per(0);
        System.out.println(maxValue);
        System.out.println(minValue);
    }

    public static void per(int i) {
        if (i == N - 1) {
            // 연산값
            int result = calculate();
            maxValue = Math.max(maxValue, result);
            minValue = Math.min(minValue, result);
            return;
        }

        for (int j = 0; j < N - 1; j++) {
            if (!selected[j]) {
                selected[j] = true;
                selectedOps[i] = ops[j];
                per(i + 1);
                selected[j] = false;
            }
        }
    }

    public static int calculate() {
        int result = arr[0];
        for (int i = 0; i < N - 1; i++) {
            switch (selectedOps[i]) {
                case 0:
                    result += arr[i + 1];
                    break;
                case 1:
                    result -= arr[i + 1];
                    break;
                case 2:
                    result *= arr[i + 1];
                    break;
                case 3:
                    result /= arr[i + 1];
                    break;
            }
        }
        return result;
    }
}
